subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/services/DefaultEnvironmentChangeTracker.kt

```diff
@@ -105,19 +119,47 @@ class DefaultEnvironmentChangeTracker(private val problemFactory: ProblemFactory
         }
 
         fun isSystemPropertyMutated(key: String): Boolean {
-            return systemPropertiesCleared || mutatedSystemProperties.containsKey(key)
+            return systemPropertiesCleared || mutatedSystemProperties[key] is SystemPropertyMutate
+        }
+
+        fun isSystemPropertyLoaded(key: String): Boolean {
+            return mutatedSystemProperties[key] is SystemPropertyLoad
+        }
+
+        fun getLoadedPropertyOldValue(key: String): Any? {
+            return (mutatedSystemProperties[key] as? SystemPropertyLoad)?.oldValue
         }
 
         fun getCachedState(): CachedEnvironmentState {
             return CachedEnvironmentState(
                 cleared = systemPropertiesCleared,
-                updates = mutatedSystemProperties.values.filterIsInstance<SystemPropertySet>(),
+                updates = mutatedSystemProperties.values
+                    .filterIsInstance<SystemPropertySet>()
+                    .filter { it !is SystemPropertyOverride },
```

Comment:
You can change `SystemPropertyOverride` to not being a subclass of `SystemPropertySet` to simplify this chain. Or see the other comment.